x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -102,6 +103,51 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
+  async rolloverAliasIfNeeded({ namespace }: { namespace?: string }) {
+    const clusterClient = await this.getClusterClient();
+    const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    const { body: simulatedRollover } = await clusterClient.indices.rollover({
+      alias,
+      dry_run: true,
+    });
+    const writeIndexMapping = await clusterClient.indices.get({
+      index: simulatedRollover.old_index,
+    });
+    const currentVersions = get(writeIndexMapping, [
+      'body',
+      simulatedRollover.old_index,
+      'mappings',
+      '_meta',
+      'versions',
+    ]);
+    const { body: simulatedIndex } = await clusterClient.indices.simulateIndexTemplate({
+      name: simulatedRollover.new_index,
+    });
+    const targetVersions = get(simulatedIndex, ['template', 'mappings', '_meta', 'versions']);
+    const componentTemplateRemoved =
+      Object.keys(currentVersions).length > Object.keys(targetVersions).length;
+    const componentTemplateUpdated = Object.entries(targetVersions).reduce(
+      (acc, [templateName, targetTemplateVersion]) => {
+        const currentTemplateVersion = get(currentVersions, templateName);
+        const templateUpdated =
+          currentTemplateVersion == null || currentTemplateVersion < Number(targetTemplateVersion);
+        return acc || templateUpdated;
+      },
+      false
+    );
```

Comment:
I think all you have to do here is check whether there's at least one component template that is newer than the previous version, or wasn't present, and I think we can use a for loop here to improve readability. Is there something we need to do when a component template is _removed_?

Reply:
I assume if someone removes a template then they intend for it not to be in the mapping anymore, so it would be important to rollover the index to remove those fields from the current write index. This ensures that any changes to the mappings will be propagated to the indices as soon as possible, whereas if we don't rollover the index on template removal then the changes wouldn't be propagated until the index rolls over naturally based on the ILM policy.

Reply:
It won't break if the fields from the removed component template stick around though. I'm a bit wary of switching between branches and continuously rolling over. Things like that can be hard to repair. That's one very plausible scenario. The other thing to consider is _if_ there is a bug here somewhere the consequences are painful because we've (imho needlessly) chosen to be pretty aggressive in rolling over.

Reply:
It will break in some cases. Certain queries like EQL and SQL can fail if mappings for a field are different across different indices, so it can be important to remove bad mappings as quickly as possible.

My issue with not rolling over aggressively is that it effectively leaves the index in kind of a "pending upgrade" state for a while after upgrade if component templates are removed. In this state, the templates have been updated but the write index won't be updated for some indeterminate amount of time - until the alias naturally rolls over due to the ILM policy. This effectively prevents us from removing any component templates and relying on the fields to no longer be mapped and will be confusing from a support perspective since the upgrade won't be truly finished until the natural rollover happens.

What's the use case for switching between branches? Switching back and forth between branches with different mappings but sharing the same ES cluster is going to break in other ways even without this template removal logic. If all component template versions on one branch are <= to the versions on the other branch, then it won't update the mappings and features may not work as expected. If some component templates are higher versioned on each branch, then it will rollover on every branch change, just as with the component removal case - this actually seems desirable to me from a dev perspective since it means the write index will have the proper mappings for the branch. Saved object migrations would have a similar issue with schema changes and switching branches - there's no downgrade capability, so once you upgrade the .kibana index on one branch there's no going back.